implementations/rust/ockam/ockam_command/src/project/show.rs

```diff
@@ -40,32 +42,113 @@ impl ShowCommand {
 }
 
-async fn run_impl(ctx: &Context, opts: CommandGlobalOpts, cmd: ShowCommand) -> miette::Result<()> {
-    let node = InMemoryNode::start(ctx, &opts.state).await?;
-    let controller = node.create_controller().await?;
+pub struct ProjectShowTui {
+    ctx: Context,
+    opts: CommandGlobalOpts,
+    project_name: Option<String>,
+}
+impl ProjectShowTui {
+    pub async fn run(
+        ctx: Context,
+        opts: CommandGlobalOpts,
+        project_name: Option<String>,
+    ) -> miette::Result<()> {
+        let tui = Self {
+            ctx,
+            opts,
+            project_name,
+        };
+        tui.show().await
+    }
+}
+
+#[ockam_core::async_trait]
+impl ShowCommandTui for ProjectShowTui {
+    const ITEM_NAME: &'static str = "projects";
+
+    fn cmd_arg_item_name(&self) -> Option<&str> {
+        self.project_name.as_deref()
+    }
+    fn terminal(&self) -> crate::Terminal<crate::TerminalStream<console::Term>> {
+        self.opts.terminal.clone()
+    }
+    async fn list_items_names(&self) -> miette::Result<Vec<String>> {
+        Ok(self.opts.state.projects.list_items_names()?)
+    }
 
-    // Lookup project
-    let id = match &opts.state.projects.get(&cmd.name) {
-        Ok(state) => state.config().id.clone(),
-        Err(_) => {
-            refresh_projects(&opts, ctx, &controller).await?;
-            opts.state.projects.get(&cmd.name)?.config().id.clone()
+    async fn get_arg_item_name_or_default(&self) -> miette::Result<String> {
+        let project = match self.cmd_arg_item_name() {
+            Some(command) => command.to_owned(),
+            None => self.opts.state.projects.default()?.name().to_owned(),
+        };
+        Ok(project)
+    }
+    async fn show_single(&self, item_name: &str) -> miette::Result<()> {
+        let node = InMemoryNode::start(&self.ctx, &self.opts.state).await?;
+        let controller = node.create_controller().await?;
+
+        let id = match &self.opts.state.projects.get(item_name) {
+            Ok(state) => state.config().id.clone(),
+            Err(_) => {
+                refresh_projects(&self.opts, &self.ctx, &controller).await?;
+                self.opts.state.projects.get(item_name)?.config().id.clone()
+            }
+        };
+
+        let project = controller.get_project(&self.ctx, id).await?;
+
+        self
+            // .opts
+            .terminal()
+            .stdout()
+            .plain(project.output()?)
+            .json(serde_json::to_string_pretty(&project).into_diagnostic()?)
+            .write_line()?;
+        self.opts
+            .state
+            .projects
+            .overwrite(&project.name, project.clone())?;
+        Ok(())
+    }
+    async fn show_multiple(&self, selected_items_names: Vec<String>) -> miette::Result<()> {
+        let node = InMemoryNode::start(&self.ctx, &self.opts.state).await?;
+        let controller = node.create_controller().await?;
+
+        for selected_item in selected_items_names.iter() {
```

Comment:
Do you prefer to create e single `String` and then print out the result to terminal once?

Reply:
You should follow the pattern from `project list`, which builds a different type of output. Check it out [here](https://github.com/build-trust/ockam/blob/331b6ebdef2717cb5c81f0b0f066d7a10412d0ff/implementations/rust/ockam/ockam_command/src/project/list.rs#L63-L67)

Reply:
Ok, so basically now is the same as project list output, with the same information (project name and space name)